Disabling the "View Unit in Studio" button to CCX courses

felipevolpone opened the pull request:
It's about this ticket: [https://openedx.atlassian.net/browse/OSPR-1383](https://openedx.atlassian.net/browse/OSPR-1383)

The idea is to remove the "View Unit in Studio" button when the user is "viewing" a CCX course, because CCX courses are not editable.

A screenshot when you're viewing a "normal" course
![screen shot 2016-08-17 at 12 25 41 am](https://cloud.githubusercontent.com/assets/1590527/17723521/27f9587e-6411-11e6-8e7b-8e50d16e465a.png)

A screenshot when you're viewing a CCX course
![screen shot 2016-08-17 at 12 25 35 am](https://cloud.githubusercontent.com/assets/1590527/17723523/2aee2ab4-6411-11e6-80a7-874a107ea46a.png)

I put the validation on the openedx/core/lib/xblock_utils.py file, but I don't know if it's a good idea, because the code from rendering the view gets coupled with the database. Instead of calling the CustomCourseForEdX model directly on the xblock_utils.py file, I could add a method on the lms/djangoapps/ccx/utils.py file that does it.

However, I don't know what is the best decision. I guess that this kind of decision it's up to you guys, even that I read the documentation about Developing at devstack, it doesn't cover this kind of thing (I also think that it shouldn't).

Also, I can't find any tests that coverage this behavior.

I added the tag (WIP) in the title, because I didn't write a test to this PR, maybe if someone give me a direction about where I could find some test as an example, it would be awesome.

Furthermore, when I run the test suite on my computer it takes more than 25 minutes to run it all, some tip about running it faster? (I used the --disable-migration option).

edx-webhook commented:
Thanks for the pull request, @felipevolpone! I've created [OSPR-1397](https://openedx.atlassian.net/browse/OSPR-1397) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



We can't start reviewing your pull request until you've submitted a     [signed contributor agreement](http://open.edx.org/sites/default/files/wysiwyg/individual-contributor-agreement.pdf)     or indicated your institutional affiliation.     If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

felipevolpone commented:
@antoviaque here is my PR :) 
I added some comments in the description, please feel free to give any feedback.

haikuginger commented on openedx/core/lib/xblock_utils.py:
```diff
@@ -302,7 +304,9 @@ def add_staff_markup(user, has_instructor_access, disable_staff_debug_info, bloc
-        if is_studio_course:
+        not_ccx_course = CustomCourseForEdX.objects.filter(course_id=course_id)
+        # will not return the View Unit in Studio button when viewing a CCX course
+        if is_studio_course and not_ccx_course:
```
I think the language around this section could use some work. `not_ccx_course` describes what we're looking for, but it doesn't give us context as to why.

haikuginger commented:
One minor comment on the code changes. In addition, it's important that all new changes also add tests. I'd suggest looking under `common/tests/acceptance`; you can also use the fixtures to help you write a test to ensure this behavior.

Once those points are addressed, I'll go ahead and review your code's behavior.

felipevolpone commented:
Hi @haikuginger, I added an assert that covers the bug fix. Also, I follow your suggestion and added a comment in the openedx/core/lib/xblock_utils.py; to explain why the button should not appear.

If something is wrong, please tell me.

haikuginger commented on openedx/core/lib/xblock_utils.py:
```diff
@@ -306,6 +306,7 @@ def add_staff_markup(user, has_instructor_access, disable_staff_debug_info, cour
 
         not_ccx_course = CustomCourseForEdX.objects.filter(course_id=course_id)
         # will not return the View Unit in Studio button when viewing a CCX course
+        # this means the user cannot edit the course on Studio
```
@felipevolpone, I'd still like to see more work on this before approving it. For example, I'd like an affirmative variable name for the variable (`is_ccx_course`) or similar, followed by something like `if is_studio_course and not is_ccx_course` for the conditional. Also, on second look, I think you might have a logic error that's masked by the variable name you're using. I would additionally take a look at the Django `QuerySet.exists()` method as a way to explicitly get a boolean from a query.

Finally, I'd like a comment describing _why_ the behavior is this way, rather than a comment simply describing what the behavior _is_. You can refer to OSPR-1383 for help in drafting this comment.

felipevolpone commented:
@haikuginger I think it's all right now, please tell me if it is not.

haikuginger commented:
Okay, I'm 👍 on this. I'm happy with the code and have manually verified that when using this branch, CCX courses do not have the "View Unit in Studio" button available that standard courses do. I don't think it needs additional documentation, as the fact that the button was available was a bug in the first place.

@bradenmacdonald, who's our upstream reviewer for this task?

bradenmacdonald commented:
@haikuginger It's an OSPR, so the Open edX team will take care of determining that.

gsong commented:
jenkins run all

bradenmacdonald commented:
@felipevolpone You've got a quality build failure for this ticket - can you please fix this?

```
18:48:25 lms/djangoapps/courseware/module_render.py (0.0%):
18:48:25     686: C0301: (line-too-long), : Line too long (122/120)
```

felipevolpone commented:
@bradenmacdonald Sure, I'll work on this as soon as possible.

felipevolpone commented:
@bradenmacdonald I think it's fine now

gsong commented:
jenkins run all

ormsbee commented on openedx/core/lib/xblock_utils.py:
```diff
@@ -302,7 +302,9 @@ def add_staff_markup(user, has_instructor_access, disable_staff_debug_info, bloc
+        # preventing users from opening ccx courses in Studio, since Studio doesn't support CCX courses
+        if is_studio_course and not is_ccx_course:
             # build edit link to unit in CMS. Can't use reverse here as lms doesn't load cms's urls.py
             edit_link = "//" + settings.CMS_BASE + '/container/' + unicode(block.location)
```
Instead of adding this toggle based on knowledge of the key, can we have CCX override `block. course_edit_method` on the blocks that it passes back so that it has `"CCX"` instead of `"Studio"`? We already have hooks for not displaying these fields for the purposes of old-style XML Modulestore courses -- might as well repurpose that logic, since it's less likely to break.

felipevolpone replied:
Hi @ormsbee, I don't know if I got it. What you're talking it's a proposal or I should follow your ideas to get this PR merged?

PS: I don't have a deep knowledge of the project, so I'll take a few days to get this done (if the change on my code is necessary).

tks

ormsbee replied:
@felipevolpone: I believe you can just change `lms/djangoapps/ccx/overrides.py:get_override_for_ccx` so that it returns `"CCX"` when the `name` passed in is `course_edit_method`. One of the main architectural goals of CCX is to keep CCX-specific knowledge from leaking out into other systems as much as possible, and to maintain the facade of "it's just another course" to querying systems. Since there is already a field toggle for, "Can we edit this in Studio?", I'd rather CCX make use of that, rather than having more CCX-specific handling code make it out to this layer.

As to whether it's required for merging -- that mostly depends on whether what I just described works, or if there are hidden complexities that I'm unaware of. I strongly prefer putting this in the CCX app instead of adding anything to `module_render()` or `add_staff_markup()`, but please let me know if this is not as straightforward as I think.

Thank you.

felipevolpone replied:
Hi @ormsbee, it has been hard to me understand where (and how) I should do the change. Looking at the `/djangoapps/ccx/overrides.py:get_override_for_ccx` method, I just tried to check the content of `block.course_edit_method`, but I got a recursive loop error.  Here is the log: http://pastebin.com/P6WH1mLi

I searched for `course_edit_method` in the project and I found this:
```python
course_edit_method = String(
        display_name=_("Course Editor"),
        help=_('Enter the method by which this course is edited ("XML" or "Studio").'),
        default="Studio",
        scope=Scope.settings,
        deprecated=True  # Deprecated because someone would not edit this value within Studio.
    )
```
in the `common/lib/xmodule/course_module.py` file. Is the idea to create another attribute just like the object above? 

I'm kind of lost here, could you help me, giving another tip or something like this?

Thank you :)

ormsbee commented:
Done with my pass.

haikuginger commented:
@felipevolpone, what's your status on this task?

felipevolpone commented:
@haikuginger sorry for the delay. I commented in the thread above.

ormsbee commented:
I'm on a commuter train, going to a conference, so I can't easily test this
out. But please try the following:

In get_override_for_ccx, in the line after block_overrides is initialized,
do:

block_overrides['course_edit_method'] = None

My guess is that will remove the unwanted links. If not, I can look at it
more closely. I'll write a better explanation when I have wifi.

On Tuesday, September 20, 2016, Felipe Volpone <notifications@github.com>
wrote:

> *@felipevolpone* commented on this pull request.
> ------------------------------
>
> In openedx/core/lib/xblock_utils.py
> <https://github.com/edx/edx-platform/pull/13252>:
>
> > @@ -302,7 +302,9 @@ def add_staff_markup(user, has_instructor_access, disable_staff_debug_info, bloc
>          # check that the course is a mongo backed Studio course before doing work
>          is_studio_course = block.course_edit_method == "Studio"
>
> -        if is_studio_course:
> +        is_ccx_course = hasattr(course_id, 'ccx')
> +        # preventing users from opening ccx courses in Studio, since Studio doesn't support CCX courses
> +        if is_studio_course and not is_ccx_course:
>              # build edit link to unit in CMS. Can't use reverse here as lms doesn't load cms's urls.py
>              edit_link = "//" + settings.CMS_BASE + '/container/' + unicode(block.location)
>
> Hi @ormsbee <https://github.com/ormsbee>, it has been hard to me
> understand where (and how) I should do the change. Looking at the
> /djangoapps/ccx/overrides.py:get_override_for_ccx method, I just tried to
> check the content of block.course_edit_method, but I got a recursive loop
> error. Here is the log: http://pastebin.com/P6WH1mLi
>
> I searched for course_edit_method in the project and I found this:
>
> course_edit_method = String(
>         display_name=_("Course Editor"),
>         help=_('Enter the method by which this course is edited ("XML" or "Studio").'),
>         default="Studio",
>         scope=Scope.settings,
>         deprecated=True  # Deprecated because someone would not edit this value within Studio.
>     )
>
> in the common/lib/xmodule/course_module.py file. Is the idea to create
> another attribute just like the object above?
>
> I'm kind of lost here, could you help me, giving another tip or something
> like this?
>
> Thank you :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/edx/edx-platform/pull/13252>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACimXk_ZPiJ0IKjGpCQMY_XCoTdmZWPks5qsJudgaJpZM4JmYO3>
> .
>

felipevolpone commented:
@ormsbee I tried, but I'm still getting the recursive error.
The code:
```python
def get_override_for_ccx(ccx, block, name, default=None):
    """
    Gets the value of the overridden field for the `ccx`.  `block` and `name`
    specify the block and the name of the field.  If the field is not
    overridden for the given ccx, returns `default`.
    """
    overrides = _get_overrides_for_ccx(ccx)

    clean_ccx_key = _clean_ccx_key(block.location)

    block_overrides = overrides.get(clean_ccx_key, {})
    block_overrides['course_edit_method'] = None

    print(block.course_edit_method)  #  the error happens here

    if name in block_overrides:
        try:
            return block.fields[name].from_json(block_overrides[name])
        except KeyError:
            return block_overrides[name]
    else:
        return default
```

ormsbee commented:
It's recursive because `get_override_for_ccx` is going to be called whenever you try to access an XBlock attribute on an XBlock in a CCX course. If you want to introspect the underlying field value, you'd want to access it via `block.fields['course_edit_method']`. But I think that if just remove the line with the print, it will do the right thing.

felipevolpone commented:
@ormsbee It worked. Now I understood what you were talking about. I "removed" my previous commits to avoid unnecessary noise in the commit log, and did all the changes in just one final commit.

Do you think it's necessary to write a comment in the line `block_overrides['course_edit_method'] = None`? I'm not sure if the impact of this line is so obvious.

ormsbee commented:
Yes, I absolutely agree that it needs a comment. Thank you!

On Wednesday, September 21, 2016, Felipe Volpone <notifications@github.com>
wrote:

> @ormsbee <https://github.com/ormsbee> It worked. Now I understood what
> you were talking about. I "removed" my previous commits to avoid
> unnecessary noise in the commit log, and did all the changes in just one
> final commit.
>
> Do you think it's necessary to write a comment in the line
> block_overrides['course_edit_method'] = None? I'm not sure if the impact
> of this line is so obvious.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/edx/edx-platform/pull/13252#issuecomment-248747472>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACimdrmNY4BKYXGaetquuvIphC_VvHZks5qsaCDgaJpZM4JmYO3>
> .
>

felipevolpone commented:
@ormsbee What do you think of this description?

```python
def get_override_for_ccx(ccx, block, name, default=None):
    # ...
    overrides = _get_overrides_for_ccx(ccx)
    clean_ccx_key = _clean_ccx_key(block.location)
    block_overrides = overrides.get(clean_ccx_key, {})

    # making course_edit_method None, a Studio link of this course
    # will not be created, so the link will not appear on the page
    block_overrides['course_edit_method'] = None

   # ...
```

ormsbee commented:
Hmmmm... maybe a little extra context with why we don't want to generate the Studio links? Like:

```python
    # Hardcode the course_edit_method to be None instead of 'Studio', so that
    # the LMS never tries to link back to Studio for these problems. CCX courses
    # can't be edited in Studio.
```

?

felipevolpone commented:
Nice! Done 👍

ormsbee commented:
👍

haikuginger commented:
jenkins run all

haikuginger commented:
jenkins run all

haikuginger commented:
@ormsbee, once I've got green tests on here and have verified the new setup manually, does anyone else from EdX need to sign off, or will this be ready to merge?

ormsbee commented:
@haikuginger: Feel free to merge at that point.

jzoldak commented:
jenkins run all

jzoldak commented:
jenkins run all

haikuginger commented:
:+1:

- [x] I tested this: created a CCX course and verified that the "View in Studio" button was not available 
- [x] I read through the code
- [x] No documentation needed; bugfix.

gsong commented:
jenkins run all
